File: benchmarks/bench.R
# R-side formula benchmark.
#
# Reads benchmarks/results/suite.json produced by bench.py and times the same
# (formula, dataset) pairs against base R, lme4, and mgcv.
#
# What each library does here:
#   - base        : stats::model.matrix(terms(formula), data)   [parametric X]
#   - lme4        : lme4::lFormula(formula, data)               [X + Zt setup]
#   - mgcv        : mgcv::gam(formula, data, fit=FALSE)         [X + smooth bases]
#
# Libraries that don't semantically handle a given fixture class are recorded
# as status="unsupported". Timing uses replicate() + Sys.time() rather than
# microbenchmark (not installed); we take min/median/max of `reps` runs.

suppressPackageStartupMessages({
  library(jsonlite)
})

script_dir <- local({
  # source()'d: sys.frame(1)$ofile; Rscript: --file=...; fallback: getwd().
  ofile <- tryCatch(sys.frame(1)$ofile, error = function(e) NULL)
  if (!is.null(ofile)) return(normalizePath(dirname(ofile)))
  cmd <- commandArgs(trailingOnly = FALSE)
  hit <- grep("^--file=", cmd, value = TRUE)
  if (length(hit) > 0) return(normalizePath(dirname(sub("^--file=", "", hit[[1]]))))
  getwd()
})
root      <- normalizePath(file.path(script_dir, ".."))
suite_js  <- file.path(root, "benchmarks", "results", "suite.json")
out_path  <- file.path(root, "benchmarks", "results", "r.csv")
data_root <- file.path(root, "datasets")

args <- commandArgs(trailingOnly = TRUE)
parse_arg <- function(key, default = NULL) {
  hit <- grep(paste0("^--", key, "="), args, value = TRUE)
  if (length(hit) == 0) return(default)
  sub(paste0("^--", key, "="), "", hit[[1]])
}
reps    <- as.integer(parse_arg("reps", "7"))
warmup  <- as.integer(parse_arg("warmup", "1"))
scales  <- as.integer(strsplit(parse_arg("scales", "1"), ",")[[1]])
libs_in <- parse_arg("libs", NULL)
kinds_in<- parse_arg("kinds", NULL)
limit   <- suppressWarnings(as.integer(parse_arg("limit", NA)))

if (!file.exists(suite_js)) stop("run bench.py first (suite.json missing)")
suite <- fromJSON(suite_js, simplifyDataFrame = FALSE)

if (!is.null(kinds_in)) {
  keep <- strsplit(kinds_in, ",")[[1]]
  suite <- Filter(function(s) s$kind %in% keep, suite)
}
if (!is.na(limit) && length(suite) > limit) suite <- suite[seq_len(limit)]

versions <- list(
  base = paste(R.version$major, R.version$minor, sep="."),
  lme4 = tryCatch(as.character(packageVersion("lme4")), error=function(e) NA),
  mgcv = tryCatch(as.character(packageVersion("mgcv")), error=function(e) NA)
)

df_cache <- new.env(hash = TRUE, parent = emptyenv())

# Parse one CSV row, returning a list of (text, was_quoted) per field. We need
# this because read.csv discards the quoting signal and will coerce "1","2","3"
# to integer — even though R's write.csv only quotes character/factor columns.
.split_csv_row <- function(line) {
  chars <- strsplit(line, "", fixed = TRUE)[[1]]
  fields <- list(); buf <- character(0)
  in_quotes <- FALSE; field_quoted <- FALSE
  n <- length(chars); i <- 1L
  while (i <= n) {
    ch <- chars[[i]]
    if (in_quotes) {
      if (ch == "\"") {
        if (i + 1L <= n && chars[[i + 1L]] == "\"") {
          buf <- c(buf, "\""); i <- i + 2L; next
        }
        in_quotes <- FALSE
      } else {
        buf <- c(buf, ch)
      }
    } else {
      if (ch == "\"") {
        in_quotes <- TRUE; field_quoted <- TRUE
      } else if (ch == ",") {
        fields[[length(fields) + 1L]] <- list(
          text = paste(buf, collapse = ""), quoted = field_quoted
        )
        buf <- character(0); field_quoted <- FALSE
      } else {
        buf <- c(buf, ch)
      }
    }
    i <- i + 1L
  }
  fields[[length(fields) + 1L]] <- list(
    text = paste(buf, collapse = ""), quoted = field_quoted
  )
  fields
}

.detect_quoted_cols <- function(path, max_rows = 50) {
  lines <- readLines(path, n = max_rows + 1L, warn = FALSE)
  if (length(lines) < 2) return(character())
  header <- vapply(.split_csv_row(lines[[1]]), function(x) x$text, character(1))
  quoted <- logical(length(header))
  for (line in lines[-1]) {
    if (nchar(line) == 0) next
    fields <- .split_csv_row(line)
    for (j in seq_len(min(length(fields), length(quoted)))) {
      if (isTRUE(fields[[j]]$quoted)) quoted[[j]] <- TRUE
    }
  }
  # Drop empty-named columns (R's write.csv row.names column). read.csv
  # synthesises "X" or treats it as row.names depending on options; either
  # way, it's never referenced by formulas.
  cols <- header[quoted]
  cols[nzchar(cols)]
}

load_df <- function(ds) {
  if (!exists(ds, envir = df_cache, inherits = FALSE)) {
    parts <- strsplit(ds, "/", fixed = TRUE)[[1]]
    path  <- file.path(data_root, parts[[1]], paste0(parts[[2]], ".csv"))
    quoted_cols <- .detect_quoted_cols(path)
    col_classes <- setNames(rep("character", length(quoted_cols)), quoted_cols)
    # stringsAsFactors=TRUE restores factor semantics; colClasses forces the
    # quoted columns through the character -> factor path even if they're all
    # digits like "1","2","3","4".
    d <- if (length(col_classes)) {
      read.csv(path, stringsAsFactors = TRUE, colClasses = col_classes)
    } else {
      read.csv(path, stringsAsFactors = TRUE)
    }
    # read.csv with colClasses="character" doesn't auto-factor; do it ourselves.
    for (col in quoted_cols) d[[col]] <- factor(d[[col]])
    assign(ds, d, envir = df_cache)
  }
  get(ds, envir = df_cache, inherits = FALSE)
}
scale_df <- function(df, k) if (k <= 1) df else df[rep(seq_len(nrow(df)), k), , drop=FALSE]

time_fn <- function(f, reps, warmup) {
  for (i in seq_len(warmup)) f()
  times <- replicate(reps, {
    t0 <- Sys.time()
    f()
    as.numeric(difftime(Sys.time(), t0, units = "secs"))
  })
  times <- sort(times)
  list(min_s = times[[1]],
       median_s = times[[ceiling(length(times)/2)]],
       max_s = times[[length(times)]])
}

# --- per-library callables --------------------------------------------------

run_base <- function(formula_str, data, kind, scope = "full") {
  f  <- stats::as.formula(formula_str)
  # `base` means model.matrix on the parametric RHS only. For mgcv formulas,
  # strip s()/te()/ti()/t2() via mgcv::interpret.gam so model.matrix doesn't
  # try to evaluate smooth constructors.
  if (kind == "mgcv") {
    pf <- mgcv::interpret.gam(f)$pf
    tt <- stats::terms(pf, data = data)
  } else {
    tt <- stats::terms(f, data = data)
  }
  stats::model.matrix(tt, data = data)
}

run_lme4 <- function(formula_str, data, kind, scope = "full") {
  f <- stats::as.formula(formula_str)
  ctl <- lme4::lmerControl(
    check.rankX        = "silent.drop.cols",
    check.nobs.vs.nlev = "ignore",
    check.nobs.vs.nRE  = "ignore",
    check.nobs.vs.rankZ= "ignore"
  )
  if (identical(scope, "parametric")) {
    # Strip the (...|group) bars so terms() sees just the parametric RHS.
    nb <- lme4::nobars(f)
    if (is.null(nb)) nb <- stats::as.formula(paste(deparse(f[[2]]), "~ 1"))
    stats::model.matrix(stats::terms(nb, data = data), data = data)
  } else {
    lme4::lFormula(f, data = data, control = ctl)
  }
}

run_mgcv <- function(formula_str, data, kind, scope = "full") {
  f <- stats::as.formula(formula_str)
  if (identical(scope, "parametric")) {
    pf <- mgcv::interpret.gam(f)$pf
    stats::model.matrix(stats::terms(pf, data = data), data = data)
  } else {
    mgcv::gam(f, data = data, fit = FALSE)
  }
}

lib_defs <- list(
  list(name = "base",  fn = run_base,  kinds = c("wr", "mgcv")),
  list(name = "lme4",  fn = run_lme4,  kinds = c("lme4")),
  list(name = "mgcv",  fn = run_mgcv,  kinds = c("mgcv"))
)

.is_data_insufficient <- function(msg) {
  pats <- c("fewer unique", "insufficient unique",
            "doesn't match margin count",
            "insufficient data", "too few")
  any(vapply(pats, function(p) grepl(p, msg, fixed = TRUE), logical(1)))
}
if (!is.null(libs_in)) {
  keep <- strsplit(libs_in, ",")[[1]]
  lib_defs <- Filter(function(l) l$name %in% keep, lib_defs)
}

dir.create(dirname(out_path), recursive = TRUE, showWarnings = FALSE)
con <- file(out_path, open = "w")
write.table(
  data.frame(library=character(), version=character(), fixture_id=character(),
             kind=character(), formula=character(), dataset=character(),
             n_rows=integer(), scale=integer(), scope=character(),
             status=character(), error=character(), n_reps=integer(),
             min_s=numeric(), median_s=numeric(), max_s=numeric()),
  con, sep = ",", row.names = FALSE, qmethod = "double"
)

total <- length(suite) * length(lib_defs) * length(scales)
i <- 0L
t0 <- Sys.time()

write_row <- function(row) {
  write.table(
    as.data.frame(row, stringsAsFactors = FALSE),
    con, sep = ",", row.names = FALSE, col.names = FALSE,
    qmethod = "double", append = TRUE
  )
  flush(con)
}

for (fx in suite) {
  df_base <- tryCatch(load_df(fx$dataset), error = function(e) NULL)
  if (is.null(df_base)) {
    for (s in scales) for (ld in lib_defs) {
      i <- i + 1L
      write_row(list(
        library=ld$name, version=as.character(if (is.null(versions[[ld$name]])) "" else versions[[ld$name]]),
        fixture_id=fx$id, kind=fx$kind, formula=fx$formula, dataset=fx$dataset,
        n_rows=0L, scale=s, status="error", error="dataset not found",
        n_reps=0L, min_s=NA_real_, median_s=NA_real_, max_s=NA_real_
      ))
    }
    next
  }
  for (s in scales) {
    df <- scale_df(df_base, s)
    for (ld in lib_defs) {
      i <- i + 1L
      status <- "ok"; err <- ""; scope <- "full"
      t <- list(min_s=NA_real_, median_s=NA_real_, max_s=NA_real_)
      if (!(fx$kind %in% ld$kinds)) {
        status <- "unsupported"
        err <- sprintf("kind=%s not in lib=%s scope", fx$kind, ld$name)
      } else {
        attempt <- function(scope_arg) tryCatch(
          list(ok = TRUE, t = time_fn(function() ld$fn(fx$formula, df, fx$kind, scope_arg), reps, warmup)),
          error = function(e) list(ok = FALSE, msg = conditionMessage(e))
        )
        res <- attempt("full")
        if (isTRUE(res$ok)) {
          t <- res$t
        } else if (.is_data_insufficient(res$msg)) {
          # Fall back to parametric: matches what the fixture oracle captured.
          res2 <- attempt("parametric")
          if (isTRUE(res2$ok)) {
            t <- res2$t; scope <- "parametric"
            err <- paste0("degraded: ", substr(res$msg, 1, 200))
          } else {
            status <- "error"; err <- substr(res2$msg, 1, 300)
          }
        } else {
          msg <- tolower(res$msg)
          status <- if (any(vapply(
            c("not supported","not implemented","unsupported","unknown",
              "no applicable","can't","cannot",
              "could not find function"),
            function(k) grepl(k, msg, fixed = TRUE), logical(1)
          ))) "unsupported" else "error"
          err <- substr(res$msg, 1, 300)
        }
      }
      row <- list(
        library=ld$name,
        version=as.character(if (is.null(versions[[ld$name]])) "" else versions[[ld$name]]),
        fixture_id=fx$id, kind=fx$kind, formula=fx$formula, dataset=fx$dataset,
        n_rows=nrow(df), scale=s, scope=scope, status=status, error=err,
        n_reps=if (status=="ok") reps else 0L,
        min_s=t$min_s, median_s=t$median_s, max_s=t$max_s
      )
      write_row(row)
      elapsed <- as.numeric(difftime(Sys.time(), t0, units="secs"))
      t_str <- if (is.na(t$min_s)) "-" else sprintf("%.2e", t$min_s)
      cat(sprintf("[%4d/%d] %-10s %-5s %-10s s=%-3d n=%-6d %-11s %s  [%5.1fs]\n",
                  i, total, ld$name, fx$kind, fx$id, s, nrow(df),
                  status, t_str, elapsed))
    }
  }
}

close(con)
cat(sprintf("\nWrote %s\n", out_path))
cat(sprintf("Versions: %s\n", toJSON(versions, auto_unbox=TRUE)))

File: benchmarks/bench.py
"""Python-side formula benchmark.

Times end-to-end "formula string + dataframe -> design matrix" for:
  - lmpy.formula   (this project)
  - formulaic
  - formulae

The formula suite is drawn from tests/fixtures/manifest.json joined with
corpus/*.yaml, filtered to fixtures with status=ok and a known dataset.

Each library is tried on every fixture; libraries that can't handle a given
formula class are recorded as `status=unsupported` with the error message,
so the reader can see coverage, not just speed.

Also writes `benchmarks/results/suite.json` so the R runner reads the exact
same list of fixtures.
"""
from __future__ import annotations

import argparse
import csv
import importlib.metadata as _md
import json
import sys
import time
import timeit
import traceback
from pathlib import Path

import polars as pl
import yaml

try:
    import pandas as pd
    _HAS_PANDAS = True
except ImportError:
    pd = None
    _HAS_PANDAS = False

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))

from lmpy.formula import (  # noqa: E402
    expand,
    materialize,
    materialize_bars,
    materialize_smooths,
    parse,
)

FIXTURES = ROOT / "tests" / "fixtures"
DATASETS = ROOT / "datasets"
CORPUS = ROOT / "corpus"
DEFAULT_OUT = ROOT / "benchmarks" / "results" / "python.csv"
SUITE_OUT = ROOT / "benchmarks" / "results" / "suite.json"


def _version(pkg: str) -> str:
    try:
        return _md.version(pkg)
    except _md.PackageNotFoundError:
        return "unknown"


VERSIONS = {
    "lmpy": _version("lmpy"),
    "formulaic": _version("formulaic"),
    "formulae": _version("formulae"),
    "polars": _version("polars"),
    "pandas": _version("pandas") if _HAS_PANDAS else "not installed",
    "numpy": _version("numpy"),
}


def load_suite() -> list[dict]:
    manifest = json.loads((FIXTURES / "manifest.json").read_text())
    ok = {e["id"]: e for e in manifest["entries"] if e.get("status") == "ok"}
    formulas: dict[str, str] = {}
    for name in ["curated.yaml", "wr.yaml", "lme4.yaml", "mgcv.yaml"]:
        doc = yaml.safe_load((CORPUS / name).read_text()) or {}
        for case in doc.get("cases") or []:
            formulas[case["id"]] = case["formula"]
    out = []
    for fx_id, e in ok.items():
        if fx_id not in formulas:
            continue
        ds = e.get("dataset")
        if not ds or "/" not in ds:
            continue
        out.append(
            {
                "id": fx_id,
                "kind": e["kind"],
                "formula": formulas[fx_id],
                "dataset": ds,
                "rows": e.get("rows"),
                "cols_X": e.get("cols_X"),
            }
        )
    out.sort(key=lambda r: (r["kind"], r["id"]))
    return out


_pd_cache: dict = {}
_pl_cache: dict[str, pl.DataFrame] = {}


def _split_csv_row(line: str) -> list[tuple[str, bool]]:
    """Parse one CSV line, returning (text, was_quoted) per field."""
    fields: list[tuple[str, bool]] = []
    buf: list[str] = []
    in_quotes = False
    field_quoted = False
    i, n = 0, len(line)
    while i < n:
        ch = line[i]
        if in_quotes:
            if ch == '"':
                if i + 1 < n and line[i + 1] == '"':
                    buf.append('"'); i += 2; continue
                in_quotes = False
            else:
                buf.append(ch)
        else:
            if ch == '"':
                in_quotes = True; field_quoted = True
            elif ch == ",":
                fields.append(("".join(buf), field_quoted))
                buf.clear(); field_quoted = False
            else:
                buf.append(ch)
        i += 1
    fields.append(("".join(buf), field_quoted))
    return fields


def _detect_quoted_columns(path, max_rows: int = 200) -> set[str]:
    """Return column names whose data values are quoted in the CSV.

    R's write.csv quotes character/factor columns and leaves numerics bare.
    pandas.read_csv and R's read.csv both ignore that signal by default and
    will happily coerce "1","2","3" to int. We recover factor semantics by
    honouring the quoting. Char-level scan (not csv.QUOTE_NONNUMERIC) so
    values like NA don't blow up the reader.
    """
    with open(path) as f:
        header_line = f.readline().rstrip("\r\n")
        if not header_line:
            return set()
        header = [t for t, _ in _split_csv_row(header_line)]
        quoted = [False] * len(header)
        for _ in range(max_rows):
            line = f.readline()
            if not line:
                break
            line = line.rstrip("\r\n")
            if not line:
                continue
            fields = _split_csv_row(line)
            for j, (_, q) in enumerate(fields):
                if j < len(quoted) and q:
                    quoted[j] = True
            if all(quoted):
                break
    # Drop empty-named columns (R's row.names convention); pandas reads those
    # as `Unnamed: 0` and they never appear in formulas.
    return {h for h, q in zip(header, quoted) if q and h}


def load_df_pd(dataset: str):
    if not _HAS_PANDAS:
        raise RuntimeError(
            "pandas is required to benchmark formulaic/formulae; "
            "install pandas or pass --libs lmpy"
        )
    if dataset not in _pd_cache:
        pkg, name = dataset.split("/", 1)
        path = DATASETS / pkg / f"{name}.csv"
        quoted = _detect_quoted_columns(path)
        dtype = {c: "string" for c in quoted} or None
        df = pd.read_csv(path, dtype=dtype)
        # R's default is stringsAsFactors=TRUE for read-in character cols;
        # mirror that so `s(x0)` on a quoted-int column hits the factor path.
        for c in quoted:
            df[c] = df[c].astype("category")
        _pd_cache[dataset] = df
    return _pd_cache[dataset]


def load_df_pl(dataset: str) -> pl.DataFrame:
    if dataset not in _pl_cache:
        pkg, name = dataset.split("/", 1)
        path = DATASETS / pkg / f"{name}.csv"
        quoted = _detect_quoted_columns(path)
        schema_overrides = {c: pl.String for c in quoted} if quoted else None
        df = pl.read_csv(path, null_values="NA", schema_overrides=schema_overrides)
        if quoted:
            df = df.with_columns([
                pl.col(c).cast(pl.Categorical) for c in quoted if c in df.columns
            ])
        _pl_cache[dataset] = df
    return _pl_cache[dataset]


def scale_pd(df, factor: int):
    if factor <= 1:
        return df
    return pd.concat([df] * factor, ignore_index=True)


def scale_pl(df: pl.DataFrame, factor: int) -> pl.DataFrame:
    if factor <= 1:
        return df
    return pl.concat([df] * factor)


# ---- per-library callables -------------------------------------------------
#
# Each callable takes (formula, df, kind, scope) where scope is "full"
# (parse+materialize+bars/smooths) or "parametric" (just the parametric X,
# matching what the fixture oracle actually captured). We fall back from full
# to parametric when the full pipeline hits data-insufficiency (e.g. a smooth
# that needs more unique covariate rows than the fixture data provides).

def _lmpy(formula_str: str, df: pl.DataFrame, kind: str, scope: str = "full"):
    f = parse(formula_str)
    ex = expand(f, list(df.columns))
    X = materialize(ex, df)
    if scope == "parametric":
        return X
    if kind == "lme4" and ex.bars:
        materialize_bars(ex, df)
    elif kind == "mgcv" and ex.smooths:
        materialize_smooths(ex, df)
    return X


def _formulaic(formula_str: str, df, kind: str, scope: str = "full"):
    from formulaic import Formula
    return Formula(formula_str).get_model_matrix(df)


def _formulae(formula_str: str, df, kind: str, scope: str = "full"):
    from formulae import design_matrices
    return design_matrices(formula_str, df)


# Each lib's native DataFrame flavor: lmpy → polars, formulaic/formulae → pandas.
LIBS: list[tuple[str, callable, str]] = [
    ("lmpy", _lmpy, "pl"),
    ("formulaic", _formulaic, "pd"),
    ("formulae", _formulae, "pd"),
]


_DATA_INSUFFICIENT = (
    "fewer unique", "insufficient unique",
    "doesn't match margin count",  # degenerate tensor dims
    "insufficient data", "too few",
)


def _is_data_insufficient(msg: str) -> bool:
    lower = msg.lower()
    return any(p in lower for p in _DATA_INSUFFICIENT)


# ---- timing ----------------------------------------------------------------

# Per-library "library can't handle this formula class" patterns. Anything
# matching is classified as `unsupported`; anything else is `error` (i.e. may
# be a real bug or a coverage gap worth filing).
_UNSUPPORTED_PATTERNS: dict[str, tuple[str, ...]] = {
    "lmpy": ("not supported", "not implemented", "notimplementederror"),
    # formulaic: bars raise FormulaSyntaxError; smooths surface as
    # FactorEvaluationError ("name 's'/'te' is not defined").
    "formulaic": (
        "not supported", "not implemented", "notimplementederror",
        "formulasyntaxerror", "operator `|`", "operator `||`",
        "factorevaluationerror", "is not defined",
    ),
    # formulae: unknown function/contrast names surface as KeyError; ^ and
    # other operators hit ScanError.
    "formulae": (
        "not supported", "not implemented", "notimplementederror",
        "keyerror:", "scanerror", "unexpected character",
    ),
}


def _classify(lib: str, msg: str) -> str:
    pats = _UNSUPPORTED_PATTERNS.get(lib, ())
    lower = msg.lower()
    return "unsupported" if any(p in lower for p in pats) else "error"


def time_one(fn, *args, reps: int, warmup: int) -> dict:
    # Warmup also surfaces any errors before we start the clock.
    for _ in range(warmup):
        fn(*args)
    t = timeit.repeat(lambda: fn(*args), number=1, repeat=reps)
    t.sort()
    return {
        "min_s": t[0],
        "median_s": t[len(t) // 2],
        "max_s": t[-1],
        "n_reps": reps,
    }


def bench_cell(lib: str, fn, fx: dict, df, reps: int, warmup: int) -> dict:
    row = {
        "library": lib,
        "version": VERSIONS.get(lib, "?"),
        "fixture_id": fx["id"],
        "kind": fx["kind"],
        "formula": fx["formula"],
        "dataset": fx["dataset"],
        "n_rows": len(df),
        "scale": None,  # filled by caller
        "scope": "full",
        "status": "ok",
        "error": "",
        "n_reps": reps,
        "min_s": "",
        "median_s": "",
        "max_s": "",
    }

    def _run(scope: str):
        return time_one(fn, fx["formula"], df, fx["kind"], scope,
                        reps=reps, warmup=warmup)

    try:
        row.update(_run("full"))
        return row
    except NotImplementedError as e:
        row["status"] = "unsupported"
        row["error"] = f"{type(e).__name__}: {e}"
        return row
    except Exception as e:
        msg = f"{type(e).__name__}: {e}".splitlines()[0][:300]
        # Data-insufficiency for a smooth/tensor: the fixture was built with
        # data that can only support parametric-oracle verification. Retry at
        # parametric scope so we still time the work the fixture oracle covers.
        if _is_data_insufficient(msg):
            try:
                row.update(_run("parametric"))
                row["scope"] = "parametric"
                row["error"] = f"degraded: {msg}"
                return row
            except Exception as e2:
                msg = f"{type(e2).__name__}: {e2}".splitlines()[0][:300]
        row["status"] = _classify(lib, msg)
        row["error"] = msg
        return row


def run(out: Path, limit: int | None, reps: int, warmup: int, scales: list[int],
        kinds: list[str] | None, libs: list[str] | None) -> None:
    suite = load_suite()
    if kinds:
        suite = [s for s in suite if s["kind"] in kinds]
    if limit:
        # Stratified: keep roughly equal counts per kind.
        from collections import defaultdict
        bykind = defaultdict(list)
        for s in suite:
            bykind[s["kind"]].append(s)
        per = max(1, limit // max(1, len(bykind)))
        suite = [s for k in bykind for s in bykind[k][:per]]

    SUITE_OUT.parent.mkdir(parents=True, exist_ok=True)
    SUITE_OUT.write_text(json.dumps(suite, indent=2))

    selected = [(name, fn, flavor) for name, fn, flavor in LIBS
                if not libs or name in libs]

    out.parent.mkdir(parents=True, exist_ok=True)
    with out.open("w", newline="") as fh:
        w = csv.DictWriter(
            fh,
            fieldnames=[
                "library", "version", "fixture_id", "kind", "formula",
                "dataset", "n_rows", "scale", "scope", "status", "error",
                "n_reps", "min_s", "median_s", "max_s",
            ],
        )
        w.writeheader()
        total = len(suite) * len(selected) * len(scales)
        i = 0
        t0 = time.time()
        needs_pd = any(flavor == "pd" for _, _, flavor in selected)
        needs_pl = any(flavor == "pl" for _, _, flavor in selected)
        for fx in suite:
            base_pd = load_df_pd(fx["dataset"]) if needs_pd else None
            base_pl = load_df_pl(fx["dataset"]) if needs_pl else None
            for s in scales:
                df_pd = scale_pd(base_pd, s) if needs_pd else None
                df_pl = scale_pl(base_pl, s) if needs_pl else None
                n_rows = len(df_pl) if df_pl is not None else len(df_pd)
                for name, fn, flavor in selected:
                    df = df_pl if flavor == "pl" else df_pd
                    i += 1
                    row = bench_cell(name, fn, fx, df, reps=reps, warmup=warmup)
                    row["scale"] = s
                    w.writerow(row)
                    fh.flush()
                    elapsed = time.time() - t0
                    tag = row["status"]
                    t_str = f"{row['min_s']:.2e}" if isinstance(row["min_s"], float) else "-"
                    print(
                        f"[{i:4d}/{total}] {name:10s} {fx['kind']:5s} "
                        f"{fx['id']:10s} s={s:<3} n={n_rows:<6} "
                        f"{tag:11s} {t_str}  [{elapsed:5.1f}s]",
                        flush=True,
                    )
    print(f"\nWrote {out}")
    print(f"Wrote suite: {SUITE_OUT}")
    print(f"Versions: {json.dumps(VERSIONS)}")


def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--out", type=Path, default=DEFAULT_OUT)
    ap.add_argument("--limit", type=int, default=None,
                    help="cap suite size (stratified across kinds)")
    ap.add_argument("--reps", type=int, default=7, help="timeit repeats")
    ap.add_argument("--warmup", type=int, default=1)
    ap.add_argument("--scales", type=str, default="1",
                    help="comma-separated row multipliers, e.g. 1,10,100")
    ap.add_argument("--kinds", type=str, default=None,
                    help="comma-separated kinds: wr,lme4,mgcv")
    ap.add_argument("--libs", type=str, default=None,
                    help="comma-separated libs: lmpy,formulaic,formulae")
    args = ap.parse_args()
    scales = [int(x) for x in args.scales.split(",") if x.strip()]
    kinds = [x.strip() for x in args.kinds.split(",")] if args.kinds else None
    libs = [x.strip() for x in args.libs.split(",")] if args.libs else None
    run(args.out, args.limit, args.reps, args.warmup, scales, kinds, libs)


if __name__ == "__main__":
    main()
